docs/content/en/docs/Design/management.md

```diff
@@ -0,0 +1,176 @@
+the scale on the message. For `m0` with scale `a` and `input1` with scale `b`,
+the result scale would be
+
+1. After multiplication: `[ab / qi]_t`.
+
+1. Before multiplication: `[a / qi * b / qi]_t` = `[ab / (qi^2)]_t`.
+
+This is messy enough. To ease the burden, we can impose additional requirement:
+Mandate a constant scale `Delta_i` for each level. This is called
+_level-specific scaling factor_. With this in mind, addition within one level
+can happen without caring about the scale, and the scale of the next level is
+also the same.
+
+1. After multiplication: `Delta_{i-1} = [Delta_i^2 / qi]_t`
+
+1. Before multiplication: `Delta_{i-1} = [Delta_i^2 / (qi^2)]_t`
+
+### BGV: Cross Level Operation
+
+With the level-specific scaling factor, one may wonder how could addition and
+multiplication of ciphertext of different levels happen as they have different
+scaling factor. This can be solved by adjusting the level and scale of the
+ciphertext of higher level.
+
+The level can be easily adjusted by dropping the extra limbs, and scale can be
+adjusted by multiplying a constant, but because multiplying a constant will
+incur additional noise, the procedure becomes the following:
+
+1. Assume the level and scale of two ciphertexts are `l1` and `l2`, `s1` and
+   `s2` respectively. WLOG assume `l1 > l2`.
+
+1. Drop `l1 - l2 - 1` limbs for the first ciphertext to make it at level
+   `l2 + 1`, if those extra limbs exist.
+
+1. Adjust scale from `s1` to `s2 * q_{l2 + 1}` by multiplying
+   `[s2 * q_{l2 + 1} / s1]_t` for the first ciphertext.
+
+1. Modulusing switching from `l2 + 1` to `l2`, the scale becomes `s2` for the
+   first ciphertext.
+
+### BGV: Implementation in HEIR
+
+In HEIR the different modulus switching policy is controlled by the pass option
+for `--secret-insert-mgmt-bgv`. The pass defaults to the "Before Multiplication"
+policy. If user wants other policy, the `after-mul` or
+`before-mul-include-first-mul` option may be used. The `mlir-to-bgv` pipeline
+option `modulus-switch-before-first-mul` corresponds to the latter option.
+
+The `secret-insert-mgmt` pass is also responsible for managing cross-level
+operations. However, as the scheme parameters has not been generated, the
+concrete scale could not be instantiated so some placeholder operations are
+inserted.
+
+After the modulus switching policy applied, user could use `generate-param-bgv`
+pass to generate scheme parameters. Optionally, user could skip this pass by
+manually providing scheme parameter as an attribute at module level.
+
+Then `populate-scale-bgv` comes into play by using the scheme parameters to
+instantiate concrete scale, and turn those placeholder operations into concrete
+multiplication operation.
+
+## CKKS
+
+CKKS is a leveled scheme where each level has a modulus `qi`. The level is
+numbered from `0` to `L` where `L` is the input level and `0` is the output
+level. CKKS ciphertext contains a scaled message $\\Delta m$ where $\\Delta$
```

Comment:
Though I am not an expert here, I had believed that the scale in CKKS was also partly used for message precision, in the sense of a fixed-point numeric scale (and given the infinite input domain for CKKS encoding). But for BGV it feels mainly like an artifact of the GHS optimization. Are there any meaningful differences in how these two kinds of "scale" are handled? I'm guessing based on this document that the answer is not for the purposes of `mgmt`, but maybe there is something about choosing an input encoding scale that is left unsaid.